dd-sdk-android-gradle-plugin/src/main/kotlin/com/datadog/gradle/plugin/DdFileUploadTask.kt

```diff
@@ -0,0 +1,327 @@
+) : DefaultTask() {
+    @get:Internal
+    internal var uploader: Uploader = OkHttpUploader()
+
+    /**
+     * The API key to use for uploading.
+     */
+    @get:Input
+    var apiKey: String = ""
+
+    private val disableGzipOption: Provider<String> =
+        providerFactory.gradleProperty(DdFileUploadTask.DISABLE_GZIP_GRADLE_PROPERTY)
+
+    /**
+     * Source of the API key set: environment, gradle property, etc.
+     */
+    @get:Input
+    var apiKeySource: ApiKeySource = ApiKeySource.NONE
+
+    /**
+     * The variant name of the application.
+     */
+    @get:Input
+    var variantName: String = ""
+
+    /**
+     * The version name of the application.
+     */
+    @get:Input
+    var versionName: String = ""
+
+    /**
+     * The version code of the application.
+     */
+    @get:Input
+    var versionCode: Int = 0
+
+    /**
+     * The service name of the application (by default, it is your app's package name).
+     */
+    @get:Input
+    var serviceName: String = ""
+
+    /**
+     * The Datadog site to upload to (one of "US1", "EU1", "US1_FED").
+     */
+    @get:Input
+    var site: String = ""
+
+    /**
+     * The url of the remote repository where the source code was deployed.
+     */
+    @get:Input
+    var remoteRepositoryUrl: String = ""
+
+    /**
+     * Build ID which will be used for mapping file matching.
+     */
+    @get:Input
+    var buildId: Provider<String> = providerFactory.provider { "" }
+
+    /**
+     * datadog-ci.json file, if found or applicable for the particular task.
+     */
+    @Optional
+    @get:InputFile
+    var datadogCiFile: File? = null
+
+    /**
+     * The sourceSet root folders.
+     */
+    @get:InputFiles
+    var sourceSetRoots: List<File> = emptyList()
+
+    /**
+     * The file containing the repository description.
+     */
+    @get:OutputFile
+    var repositoryFile: File = File("")
+
+    init {
+        group = DdAndroidGradlePlugin.DATADOG_TASK_GROUP
+        // it is never up-to-date, because request may fail
+        outputs.upToDateWhen { false }
+    }
+
+    @Internal
+    internal abstract fun getFilesList(): List<Uploader.UploadFileInfo>
+
+    /**
+     * Uploads the files retrieved from `getFilesList` to Datadog.
+     */
+    @TaskAction
+    @Suppress("TooGenericExceptionCaught")
+    fun applyTask() {
+        datadogCiFile?.let {
+            applyDatadogCiConfig(it)
+        }
+        applySiteFromEnvironment()
+        validateConfiguration()
+
+        check(!(apiKey.contains("\"") || apiKey.contains("'"))) {
+            INVALID_API_KEY_FORMAT_ERROR
+        }
+
+        check(buildId.isPresent && buildId.get().isNotEmpty()) {
+            MISSING_BUILD_ID_ERROR
+        }
+
+        val mappingFiles = getFilesList()
+        if (mappingFiles.isEmpty()) return
+
+        val repositories = repositoryDetector?.detectRepositories(
+            sourceSetRoots,
+            remoteRepositoryUrl
+        ).orEmpty()
```

Comment:
now once `repositoryDetector` is not null, this can be:

```suggestion
        val repositories = repositoryDetector.detectRepositories(
            sourceSetRoots,
            remoteRepositoryUrl
        )
```